File: src/deepthrottle/algorithms/sliding_window_counter.py
import time


class SlidingWindowCounter:
    """
    A sliding window counter that tracks the number of requests in a given window size.
    - max_requests: that is the maximum number of requests allowed in the window
    - window_size_seconds: the size of the window in seconds
    """
    def __init__ (self, max_requests: int, window_size_seconds: float):
        self.max_requests: int = max_requests
        self.window_size: float = window_size_seconds
        self.previous_window_requests: int = 0
        self.current_window_requests: int = 0
        self.current_window_start: float = time.monotonic()

    def _advance_window(self, now :float) -> None:
        elapsed = now - self.current_window_start
        
        if elapsed >= self.window_size:                                                                                                                                                                      
            windows_passed = int(elapsed // self.window_size)                                                                                                                                                
            if windows_passed == 1:                                                                                                                                                                          
                self.previous_window_requests = self.current_window_requests                                                                                                                                       
            else:                                                                                                                                                                                            
                self.previous_window_requests = 0                                                                                                                                                               
            self.current_window_requests = 0                                                                                                                                                                    
            self.current_window_start += windows_passed * self.window_size

    def add_request(self, cost: int = 1) -> tuple[int, bool, float]:                                                                                                                                         
        now = time.monotonic()                                                                                                                                                                               
        self._advance_window(now)                                                                                                                                                                            
                                                                                                                                                                                                                
        current_elapsed = now - self.current_window_start                                                                                                                                                    
        weight_prev = max(0.0, 1.0 - (current_elapsed / self.window_size))                                                                                                                                   
        estimated_requests = self.current_window_requests + (weight_prev * self.previous_window_requests)                                                                                                          
                                                                                                                                                                                                                
        if estimated_requests + cost <= self.max_requests:                                                                                                                                                   
            self.current_window_requests += cost                                                                                                                                                                
            remaining = max(0, int(self.max_requests - (estimated_requests + cost)))                                                                                                                         
            return remaining, True, 0.0                                                                                                                                                                      
                                                                                                                                                                                                                
        remaining = max(0, int(self.max_requests - estimated_requests))                                                                                                                                      
        wait_time = self.window_size - current_elapsed                                                                                                                                                       
        return remaining, False, wait_time                                                                                                                                                                   
                                                                                                                                                                                                                
    def get_state(self) -> dict[str, float | str]:                                                                                                                                                           
        now = time.monotonic()                                                                                                                                                                               
        self._advance_window(now)                                                                                                                                                                            
        current_elapsed = now - self.current_window_start                                                                                                                                                    
        weight_prev = max(0.0, 1.0 - (current_elapsed / self.window_size))                                                                                                                                   
        estimated = self.current_window_requests + (weight_prev * self.previous_window_requests)                                                                                                                   
        return {                                                                                                                                                                                             
            "Name": "SlidingWindowCounter",                                                                                                                                                                  
            "Window Size": self.window_size,                                                                                                                                                                 
            "Previous Windows Count": self.previous_window_requests,                                                                                                                                            
            "Current Window Count": self.current_window_requests,                                                                                                                                               
            "Estimated Count": round(estimated, 2),                                                                                                                                                          
            "Remaining Requests": max(0, int(self.max_requests - estimated)),                                                                                                                                
        }

File: src/deepthrottle/storage/lua/sliding_window_counter.lua
-- Sliding Window Counter Atomic Lua Script
-- KEYS[1] = limiter key (e.g. 'deepthrottle:limiter:user-123')
-- ARGV[1] = max requests allowed in the window
-- ARGV[2] = window size in seconds
-- ARGV[3] = requested cost / requests count (default: 1)
-- ARGV[4] = current timestamp (seconds as float)
-- ARGV[5] = key TTL (seconds)

local key = KEYS[1]
local max_requests = tonumber(ARGV[1])
local window_size = tonumber(ARGV[2])
local requested = tonumber(ARGV[3])
local now = tonumber(ARGV[4])
local ttl = tonumber(ARGV[5]) or math.ceil(window_size * 2)

-- HINT 1: Fetch current window counts and window start time from Redis Hash
local data = redis.call('HMGET', key, 'curr_count', 'prev_count', 'window_start')
local curr_count = tonumber(data[1]) or 0
local prev_count = tonumber(data[2]) or 0
local window_start = tonumber(data[3]) or now

-- HINT 2: Check if window has rolled over
local elapsed = now - window_start
if elapsed >= window_size then
    local windows_passed = math.floor(elapsed / window_size)
    if windows_passed == 1 then
        prev_count = curr_count
    else
        prev_count = 0
    end
    curr_count = 0
    window_start = window_start + (windows_passed * window_size)
    elapsed = now - window_start
end

-- HINT 3: Calculate weighted request count
local weight_prev = math.max(0.0, 1.0 - (elapsed / window_size))
local estimated = curr_count + (weight_prev * prev_count)

local allowed = 0
local wait_time = 0

if estimated + requested <= max_requests then
    curr_count = curr_count + requested
    allowed = 1
else
    wait_time = math.max(0.0, window_size - elapsed)
end

local remaining = math.max(0, math.floor(max_requests - (estimated + (allowed == 1 and requested or 0))))

-- HINT 4: Save state and refresh TTL
redis.call('HSET', key, 'curr_count', curr_count, 'prev_count', prev_count, 'window_start', window_start)
redis.call('EXPIRE', key, ttl)

-- Return: {allowed (1 or 0), remaining, wait_time}
return {allowed, remaining, wait_time}